hack/gha-reversemap.sh

```diff
@@ -0,0 +1,260 @@
+Harmonise '$GITHUB_WORKFLOWS_PATH' and the reversemap '$REVERSEMAP_FILE'
+
+Usage:
+    $0  <operation>
+
+Example:
+    # Update actions/checkout to its latest release
+    $0  update actions/checkout
+
+Operations:
+    apply                 -   Apply the commit sha of every actions listed in the reversemap as the version to use in all github workflows
+    update <action-ref>   -   Update the version of the given action reference to its latest release within the reverse map (sha, tag, urls)
+    sync                  -   Sync the reverse map values (sha, tag, urls) from the actions used in all github workflows
+
+Parameters:
+    <action-ref>:   {gh_owner}/{gh_repo}
+EOF
+}
+
+# Log information
+_loginfo() {
+    echo -e "$(date -Iseconds);INFO;$1"
+}
+
+# Exit with an error
+_error_exit() {
+    echo -e "$(date -Iseconds);ERROR;$2" >&2
+    exit "$1"
+}
+
+# Internal - indicates proper return of value within a function
+_return() {
+    echo "$1"
+}
+
+# Function to check if yq is installed and has the required version
+_check_yq_version() {
+    if command -v yq >/dev/null 2>&1; then
+        INSTALLED_VERSION=$(yq --version 2>/dev/null)
+        if ! [[ "$INSTALLED_VERSION" =~ $YQ_REQUIRED_VERSION ]]; then
+            _error_exit $ERR_YQ_NOT_INSTALLED "yq is installed but the version is $INSTALLED_VERSION. Required version is $YQ_REQUIRED_VERSION."
+        fi
+    else
+        _error_exit $ERR_YQ_NOT_INSTALLED "yq is not installed."
+    fi
+}
+
+# Function to download and install yq
+_install_yq() {
+    os=$(uname | tr '[:upper:]' '[:lower:]')
+    arch=$(uname -m)
+
+    if [[ "$arch" == "x86_64" ]]; then
+        arch="amd64"
+    elif [[ "$arch" == "aarch64" ]]; then
+        arch="arm64"
+    elif [[ "$arch" == "arm64" ]]; then
+        arch="arm64"
+    else
+        _error_exit $ERR_ARCH_UNSUPPORTED "Unsupported architecture: $arch"
+    fi
+
+    yq_binary="yq_${os}_${arch}"
+    yq_binary_url="https://github.com/mikefarah/yq/releases/download/${YQ_REQUIRED_VERSION}/${yq_binary}.tar.gz"
+
+    _loginfo "Downloading yq $YQ_REQUIRED_VERSION for $os/$arch ..."
+    curl -L -o "/tmp/${yq_binary}.tar.gz" "$yq_binary_url"
+
+    if [[ $? -ne 0 ]]; then
+        _error_exit $ERR_YQ_DOWNLOAD_FAILED "Failed to download yq."
+    fi
+
+    tar xzf "/tmp/${yq_binary}.tar.gz" -C /tmp
+    sudo chmod +x "/tmp/${yq_binary}"
+    sudo mv "/tmp/${yq_binary}" /usr/local/bin/yq
+
+    echo "yq $YQ_REQUIRED_VERSION installed successfully."
+}
+
+# Get commitsha from an action ref upstream
+_fetch_sha_from_upstream_ref() {
+    action_ref=$1
+    tag_or_branch=$2
+    action_ref_safe=$(echo "$action_ref" | cut -d '/' -f 1,2)
+    API_GITHUB_BRANCH="https://api.github.com/repos/${action_ref_safe}/git/refs/heads/${tag_or_branch}"
+    API_GITHUB_TAG="https://api.github.com/repos/${action_ref_safe}/git/refs/tags/${tag_or_branch}"
+    HTTP_STATUS=$(curl -o /dev/null -s -w "%{http_code}\n" -H "Authorization: Bearer $GITHUB_TOKEN" "$API_GITHUB_TAG")
```

Comment:
FYI, you could reduce the number of `curl` invocations here from 3 to 2 by combining `-o /tmp/$$.json` and `-w %{http_code}` in one.